docs/proposals/001-k8s-status.md

```diff
@@ -0,0 +1,432 @@
+
+You can do this by describing user scenarios, and how this feature helps them.
+You can also provide examples of how this feature may be used.
+-->
+
+The current `k8s status` command returns information about the control plane
+nodes, and then mimics the output of the `k8s get` command. This is not useful,
+as this is by definition the "spec" of what the built-in features should be, not
+their status.
+
+If anything happens and deploying a built-in feature fails, then `k8s status`
+does not present this information in any way to the user. Further, the only way
+to debug this is to look for logs in the `k8sd` service, which users might not
+be familiar with.
+
+Instead, we want the `k8s status` command to return meaningful information about
+the current _deployment status_ of each feature. We want to replace the
+following output (taking the `dns` feature as an example):
+
+```bash
+$ k8s status
+...
+dns:
+  enabled: true                         # these fields comes from the cluster
+  service-ip: 10.152.183.10             # configuration. enabled=true does not
+  cluster-domain: cluster.local         # mean that 'dns' is actually available
+```
+
+With something like this:
+
+```bash
+$ k8s status
+...
+dns:
+  enabled: true/false                   # true if successful, false otherwise
+  message: CoreDNS deployed
+```
+
+## User facing changes
+<!--
+This section MUST describe any user-facing changes that this feature brings, if
+any. If an API change is required, the affected endpoints MUST be mentioned. If
+the output of any k8s command changes, the difference MUST be mentioned, with a
+clear example of "before" and "after".
+-->
+
+### `k8s status` command
+
+The output format of `k8s status` will change.
+
+#### Old output
+
+```bash
+$ k8s status
+status: ready
+high-availability: no
+datastore:
+  type: k8s-dqlite
+  voter-nodes:
+    - 10.87.24.119:6400
+  standby-nodes: none
+  spare-nodes: none
+network:
+  enabled: true
+dns:
+  enabled: true
+  cluster-domain: cluster.local
+  service-ip: 10.152.183.104
+  upstream-nameservers:
+  - /etc/resolv.conf
+ingress:
+  enabled: false
+  default-tls-secret: ""
+  enable-proxy-protocol: false
+load-balancer:
+  enabled: false
+  cidrs: []
+  l2-mode: false
+  l2-interfaces: []
+  bgp-mode: false
+  bgp-local-asn: 0
+  bgp-peer-address: ""
+  bgp-peer-asn: 0
+  bgp-peer-port: 0
+local-storage:
+  enabled: true
+  local-path: /var/snap/k8s/common/rawfile-storage
+  reclaim-policy: Delete
+  default: true
+gateway:
+  enabled: true
+```
+
+#### New output
+
+With `--output-format=plain`, a simple overview is shown, with `yaml` or `json`
+more fields are shown:
+
+```bash
+$ k8s status
+status: ready
+high-availability: no
+datastore:
+  type: k8s-dqlite
+  voter-nodes:
+    - 10.87.24.119:6400
+  standby-nodes: none
+  spare-nodes: none
+network:
+  enabled: true
+  message: Cilium is deployed
+dns:
+  enabled: true
+  message: CoreDNS is deployed
+ingress:
```

Comment:
I think the status of most of these components could easily be reduced to one line if we wanted to be brief

Reply:
should we do something like? (for the plain output)
```
network: enabled/disabled
ingress: enabled/disabled
```

Reply:
my fear is that this prevents us from expanding on this output in the future, if more info is needed for each feature.

Reply:
I was thinking something that gives some relevant info where useful. Just as an example:

e.g.
```
 sudo k8s status
cluster status:	   ready at 10.152.183.1
connected nodes:   10.152.183.16, 10.152.183.18,...[6 other nodes]
high-availability: no
datastore:         k8s-dqlite running
network:           enabled
dns: 		   enabled for cluster.local
ingress:           disabled
load-balancer:     disabled
local-storage:     enabled at /var/snap/k8s/common/rawfile-storage
gateway:           enabled
```
I don't really see the point of 9 lines of yaml config for the load balancer when it isn't even on.  If you want the full dump of everything that can be an option, and we should tell people that in a line at the end.

I am coming from a position of "what would the user find useful", not "what can we tell them" and I think the starting point should be that - why are you running 'status'?

Reply:
this reads like a great suggestion to me, more than happy to adjust accordingly. then it's up to the implementation to pick the level of detail for each component

Reply:
By the way, a few more thoughts on this: This subject touches on another peeve of mine about the current `k8s status` output, where the k8sd nodes appear as the datastore nodes but that is not really true (the microcluster endponts are not the cluster datastore, and are probably not even useful info)

I greatly see the value of adjusting `k8s status` to look like this, but the proposal started with the aim of iterating on it, one step at a time. as you said, there is no reason to show 9 lines of load balancer configs when people run `k8s status`, so we can start improving from this, in small iterations.

But I like your suggestion, it will be more involving for the implementation, but it will make `k8s status` a lot more useful than it currently is. Let me update the document accordingly, thanks!